src/sentry/static/sentry/app/views/organizationHealth/util/healthRequest.jsx

```diff
@@ -129,25 +267,131 @@ class HealthRequestWithParams extends React.Component {
     });
   };
 
-  transformData = () => {
-    let {timeseries, tag} = this.props;
-    let {data} = this.state;
+  /**
+   * Transforms query response into a non-timeseries data to be used in a chart
+   */
+  transformNonTimeSeriesData = (data, tag) =>
+    data.map(({[tag]: tagObject, count}) => [this.getCategory(tagObject), count]);
+
+  transformData = (data, isTimeseries) => {
+    const {tag} = this.props;
     if (!data) return null;
 
-    return timeseries
-      ? this.transformTimeseriesData()
-      : data.map(({[tag]: name, count}) => [name, count]);
+    return isTimeseries
+      ? this.transformTimeseriesData(data, tag)
+      : this.transformNonTimeSeriesData(data, tag);
   };
 
+  processData({data, totals} = {}, isTimeseries) {
+    const {
+      tag,
+      includeTransformedData,
+      includePercentages,
+      includeTimeAggregation,
+      includeTop,
+    } = this.props;
+    const shouldIncludePercentages = includePercentages && includeTop && !isTimeseries;
+    const {current, previous} = this.getData(data, isTimeseries);
+    const transformedData =
+      includeTransformedData || shouldIncludePercentages
+        ? this.transformData(current, isTimeseries)
+        : null;
+
+    const percentageMap =
+      shouldIncludePercentages &&
+      totals &&
+      this.transformSeriesPercentageMap(transformedData, totals.count);
+
+    const dataWithPercentages =
+      shouldIncludePercentages && current
+        ? current.map(({count, lastCount, [tag]: tagObject}) => {
+            const name = this.getCategory(tagObject);
+
+            return {
+              count,
+              lastCount,
+              name,
+              percentage: percentageMap.get(name),
+            };
+          })
+        : null;
+
+    const previousData =
+      isTimeseries && includeTransformedData
+        ? this.transformPreviousPeriodData(current, previous)
+        : null;
+
+    const timeAggregatedData =
+      isTimeseries && includeTimeAggregation
+        ? this.transformAggregatedTimeseries(current, includeTimeAggregation)
+        : null;
+
+    return {
+      data: transformedData,
+      allData: data,
+      originalData: current,
+      totals,
+      originalPreviousData: previous,
+      previousData,
+      timeAggregatedData,
+      dataWithPercentages,
+    };
+  }
+
   render() {
-    let {children} = this.props;
-    let {data} = this.state;
+    const {children, tag, showLoading, ...props} = this.props;
+
+    const {tagData, timeseriesData} = this.state;
+
+    // Is "loading" if data is null
+    const loading = tagData === null && timeseriesData === null;
+
+    if (showLoading && loading) {
+      return <LoadingIndicator />;
+    }
+
+    const {
+      data: transformedTagData,
+      allData: allTagData,
+      originalData: originalTagData,
+      totals: tagTotals,
+      dataWithPercentages: tagDataWithPercentages,
+    } =
+      (tagData && this.processData(tagData)) || {};
```

Comment:
I think ideally we would never need unprocessed data, but since our visualizations are not really finalized yet, sometimes we may want it just to get things working.

I think the fetching function should only be responsible for making and returning the request, that allows other components that maybe don't have a chart or table to ingest it however it needs. This allows `HealthRequest` to be responsible for any data transforms and is generally consumed by charts/table components.

Reply:
Makes sense for now, I think it could be done together in the code in some way in future (even if not exactly part of fetch function)